examples/remix-with-typescript/app/entry.server.tsx

```diff
@@ -0,0 +1,57 @@
+import * as React from 'react';
+import { renderToString } from 'react-dom/server';
+import { RemixServer } from 'remix';
+import type { EntryContext } from 'remix';
+
+import createEmotionCache from './src/createEmotionCache';
+import theme from './src/theme';
+import StylesContext from './src/StylesContext';
+
+import CssBaseline from '@mui/material/CssBaseline';
+import { ThemeProvider } from '@mui/material/styles';
+import { CacheProvider } from '@emotion/react';
+import createEmotionServer from '@emotion/server/create-instance';
+
+export default function handleRequest(
+  request: Request,
+  responseStatusCode: number,
+  responseHeaders: Headers,
+  remixContext: EntryContext,
+) {
+  const cache = createEmotionCache();
+  const { extractCriticalToChunks } = createEmotionServer(cache);
+
+  const MuiRemixServer = () => (
+    <CacheProvider value={cache}>
+      <ThemeProvider theme={theme}>
+        {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
+        <CssBaseline />
+        <RemixServer context={remixContext} url={request.url} />
+      </ThemeProvider>
+    </CacheProvider>
+  );
+
+  // Render the component to a string.
+  const html = renderToString(
+    <StylesContext.Provider value={null}>
+      <MuiRemixServer />
+    </StylesContext.Provider>,
+  );
+
+  // Grab the CSS from emotion
+  const emotionChunks = extractCriticalToChunks(html);
+
+  // Re-render including the extracted css.
+  const markup = renderToString(
```

Comment:
I have just realized that we render twice. In https://remix.run/docs/en/v1/guides/styling

> It's unlikely this will effect performance in a significant way, React is pretty fast.

I don't buy it, not for one second. React is slow on the server side. At least, it's my personal experience at Onepixel, indexing about 10m pages with Next.js. Most of the time spent to get the first byte was on rendering the React tree (react apollo was half of this time). Looking back then, I can't imagine that it would have flied to render twice.

Could we look into the feasibility to render only once? :)

Reply:
Adding it to my Todo list

Reply:
@mnajdova I am wondering if this can be extended to render head+body in the first pass and only head in the second: https://gist.github.com/alexkuz/d91c9ce1380f4b528ebf262ad0e4f959

that would be so far the best way of using css-in-js with remix - until it gets better support for build time css-in-js.

Reply:
Hey @jurokapsiar :) That's a good idea. I plan to look into this next week when there is not much activity on the repo. I will open a PR with my findings.

Reply:
I had some time today, have created https://github.com/mui-org/material-ui/pull/30366 

I am basically inserting the style tags after an insertion point meta element that is added by the `root.tsx`. As far as I could test it works, but I may be missing something. It feels a bit hacky to parse the HTML string, but I don't have a better idea.